pkg/streamer/a11y_infer.go

```diff
@@ -0,0 +1,118 @@
+package streamer
+
+import (
+	"github.com/readium/go-toolkit/pkg/internal/extensions"
+	"github.com/readium/go-toolkit/pkg/manifest"
+	"github.com/readium/go-toolkit/pkg/mediatype"
+)
+
+func inferA11yMetadataFromManifest(mf manifest.Manifest) *manifest.A11y {
+	var a11y manifest.A11y
+	if mf.Metadata.Accessibility != nil {
+		a11y = *mf.Metadata.Accessibility
+	} else {
+		a11y = manifest.NewA11y()
+	}
+
+	accessModes := a11y.AccessModes
+	accessModesSufficient := a11y.AccessModesSufficient
+	features := a11y.Features
+
+	addFeature := func(f manifest.A11yFeature) {
+		if !extensions.Contains(features, f) {
+			features = append(features, f)
+		}
+	}
+
+	allResources := append(mf.ReadingOrder, mf.Resources...)
+
+	if len(accessModes) == 0 {
+		for _, link := range allResources {
+			if link.MediaType().IsAudio() || link.MediaType().IsVideo() {
+				accessModes = append(accessModes, manifest.A11yAccessModeAuditory)
+				break
+			}
+		}
+
+		for _, link := range allResources {
+			if link.MediaType().IsBitmap() || link.MediaType().IsVideo() {
+				accessModes = append(accessModes, manifest.A11yAccessModeVisual)
+				break
+			}
+		}
+	}
+
+	if len(accessModesSufficient) == 0 {
+		setTextual := false
+
+		for _, profile := range a11y.ConformsTo {
+			if profile == manifest.EPUBA11y10WCAG20A ||
+				profile == manifest.EPUBA11y10WCAG20AA ||
+				profile == manifest.EPUBA11y10WCAG20AAA {
+				setTextual = true
+				break
+			}
+		}
+
+		if !setTextual {
+			setTextual = true
+			for _, link := range allResources {
+				mt := link.MediaType()
+				if mt.IsAudio() ||
+					mt.IsVideo() ||
+					(mt.IsBitmap() && !extensions.Contains(link.Rels, "cover")) ||
+					mt.Matches(&mediatype.PDF) {
```

Comment:
@chocolatkey Do you remember why we pass `MediaType` pointers instead of values to `Matches()`? This doesn't feel right as media types are value-based.

Reply:
@mickael-menu That was quite a while ago and I've forgotten. But I'm going to guess it's because a nil value is an acceptable possibility to provide as a potential match (which will of course match false) and you don't have to nil check.

But then I thought a bit more, and remembered if something looks strange in this way it's probably because I was trying to stay close to what the kotlin toolkit did, and sure enough we have a [nullable MediaType in there](https://github.com/readium/kotlin-toolkit/blob/main/readium/shared/src/main/java/org/readium/r2/shared/util/mediatype/MediaType.kt#L196), so that's almost certainly why:
![image](https://user-images.githubusercontent.com/13407784/213671693-b20326b8-1eb4-4490-9db3-69bfa43895c3.png)
And I was/am replicating the exact logic in the function `Matches` wraps, the contains function which also lets you pass in null values it just ignores
![image](https://user-images.githubusercontent.com/13407784/213671965-4fe1fa40-238c-4841-aecf-30619832e58b.png)